field/TinyLang/Field/Raw/Parser.hs

```diff
@@ -0,0 +1,360 @@
+{-|
+= Raw Parser
+
+A simple parser for language involving booleans, field elements, and
+vectors.
+
+Given that the language is simple enough we are mixing lexical
+analysis and grammar a bit.  To avoid any ambiguities between lexical
+analysis and grammar we will explicitly mark all lexical tokens.
+
+Please note that expressions are the designated start production for
+the parser.
+
+We are using the following convention
+
+* /tokens/ are enclosed in parentheses @""@.
+* /regular expressions/ are enclosed in square brackets @[]@.
+
+== Literals
+
+Literals are defined as follows
+
+@
+bool-literal ::= "T" | "F"
+vec-literal  ::= "{" (bool-literal ("," bool-literal)*)? "}"
+int-literal  ::= [0-9]+
+@
+
+== Keywords
+
+Here is a list of keywords used in the language.
+
+@
+keyword ::=
+    "and"
+    "assert"
+    "do"
+    "else"
+    "end"
+    "for"
+    "if"
+    "inv"
+    "let"
+    "neq0"
+    "not"
+    "or"
+    "then"
+    "to"
+    "unpack"
+    "xor"
+@
+
+== Identifiers
+
+To avoid a syntactic clash with /bool-literals/, identifiers for field
+variables start with a lower-case letter.  Identifiers for boolean
+variables are prefiex by @?@ and identifiers for vector variables are
+prefixed by @#@.  We will check if an identifier is not a keyword.
+
+@
+ident ::= ( "?" | "#" ) [a-z] ([a-z0-9_'])*
+@
+
+== Constants
+
+At the moment constants consist of literals only.
+
+@
+const ::=
+    bool-literal
+    int-literal
+    vec-literal
+@
+
+== Expressions
+
+As mentioned above @expr@ is the designated start production for the
+parser.
+
+@
+expr ::=
+    const
+    "(" expr ")"
+    ident
+    expr infix-op expr
+    prefix-op expr
+    expr "[" expr "]"
+    statement ";" expr
+    "if" expr "then" expr "else" expr
+
+infix-op ::=
+    "and"
+    "or"
+    "xor"
+    "=="
+    "<="
+    "<"
+    ">="
+    ">"
+    "+"
+    "-"
+    "*"
+    "/"
+
+prefix-op ::=
+    "not"
+    "neq0"
+    "neg"
+    "inv"
+    "unpack"
+@
+
+== Statement
+
+Statements are a bit odd at the moment, as they are neither
+expressions nor the usual statements.
+
+@
+statement ::=
+    "let" var "=" expr
+    "assert" expr
+    "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+    (statement (";" statement)*)?
```

Comment:
@effectfully, I have a question about parsing `statements`:  should the grammar permit empty statements or empty expressions?

First of all, I think you already mentioned that the grammar is slightly akward, with statements being the other way round, and currently, the smallest program with a statement is
```
assert T; T
```
For for loops the original grammar from `Parser.hs` permits `statements` to be empty, we should be able to write
```
for x = 0 to 10 do end; T
```
but the AST does not permit the above.

Theoretically, we can permit empty loops.  Should I edit the grammar to permit them?

Also, neither grammar nor ast permit the following program:
```
assert T;
```

Reply:
> I have a question about parsing statements: should the grammar permit empty statements or empty expressions?

We don't want any empty expressions. We may want empty statements in the raw AST just for the ease of targeting and optimizing it (fewer invariants to preserve). We'll then need to remove them when mapping the raw AST to the typed one.

Permitting empty loops may be convenient for program generation (again, fewer invariants to worry about).

So yeah, I think should we allow empty loops. Please edit the grammar.

Oh, I totally forgot. Could you also add `for` loops to the AST and parse them directly instead of unrolling them on the fly like it was done previously? Like

```haskell
data Statement f
    = <...>
    | EFor Var (Expr f) (Expr f) [Statement f]
```

(a loop variable, a lower bound, an upper bound, a list of statements). Feel free to either do that here or in a separate PR depending on what is more convenient.

> Also, neither grammar nor ast permit the following program:
>
> `assert T;`

Yeah, that is something that we'll fix later.